error.go

```diff
@@ -1,37 +1,210 @@
 package enigma
 
 import (
+	"encoding/json"
 	"fmt"
+	"reflect"
+	"runtime/debug"
+	"strings"
 )
 
 type (
-	// Error extends the built in error type with extra error information provided by the Qlik Associative Engine
+	// Error extends the built in error type with extra error information provided by the Qlik Associative Engine.
+	// Code returns the provided error code from engine.
+	// Parameter gives some context for the error.
+	// Message returns the error message from Engine.
+	// CallParams returns contains the parameters used when calling engine if any.
+	// Trace returns the stack trace of the error, omitting some superfluous lines.
+	// Caller returns the first call in the stack, containing function and file.
 	Error interface {
 		error
 		Code() int
 		Parameter() string
 		Message() string
+		CallParams() string
+		Trace() string
+		Caller() string
 	}
 
+	//qixError contains the stacktrace leading to the error and may contain the parameters used when
+	//the error occurred as well as the underlying error informat from engine.
 	qixError struct {
-		ErrorCode      int    `json:"code"`
-		ErrorParameter string `json:"parameter"`
-		ErrorMessage   string `json:"message"`
+		code       int
+		parameter  string
+		message    string
+		stack      []byte
+		callParams []interface{}
 	}
 )
 
-func (err *qixError) Error() string {
-	return fmt.Sprintf("%s: %s (%d)", err.ErrorParameter, err.ErrorMessage, err.ErrorCode)
+func (e *qixError) Error() string {
+	return fmt.Sprintf("%s: %s (%d)", e.parameter, e.message, e.code)
 }
 
-func (err *qixError) Code() int {
-	return err.ErrorCode
+func (e *qixError) Code() int {
+	return e.code
 }
 
-func (err *qixError) Parameter() string {
-	return err.ErrorParameter
+func (e *qixError) Parameter() string {
+	return e.parameter
 }
 
-func (err *qixError) Message() string {
-	return err.ErrorMessage
+func (e *qixError) Message() string {
+	return e.message
+}
+
+func (e *qixError) UnmarshalJSON(text []byte) error {
+	m := map[string]interface{}{}
+	// We don't really want to build our own unmarshaler.
+	// We just want to tack on the stack during the unmarshalling.
+	if err := json.Unmarshal(text, &m); err != nil {
+		return err
+	}
+	e.unmarshalMap(m)
+	e.stack = debug.Stack()
+	return nil
+}
+
+// Convert the result of unmarshalling into a qixError
+func (e *qixError) unmarshalMap(m map[string]interface{}) error {
+	e.parameter = m["parameter"].(string)
+	e.message = m["message"].(string)
+	// For some reason what looks and should be an int is a float64 so we
+	defer func() { // get a panic that we have to handle with this defer.
+		if r := recover(); r != nil {
+			f := m["code"].(float64)
```

Comment:
f, ok := m["code"].(float64) instead of panic?

Reply:
Derp. Good catch, definitely no need to panic!
I wonder if we can be sure that the code will always be a `float64` though... I'll use a type switch instead for now.